transaction/context.go

```diff
@@ -0,0 +1,180 @@
+	tx, err := new()
+	if err != nil {
+		return nil, err
+	}
+
+	return context.WithValue(ctx, transactionKey, tx), nil
+}
+
+// TxContext creates a new transaction context from context.Background()
+func TxContext() (ctx context.Context, err error) {
+	return NewContext(context.Background())
+}
+
+// FromContext Retrieves the transaction from the context.
+func FromContext(ctx context.Context) (*sql.Tx, error) {
+	transaction, err := getTransaction(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return transaction.tx, err
+}
+
+func getTransaction(ctx context.Context) (*transaction, error) {
+	transaction, ok := ctx.Value(transactionKey).(*transaction)
+	if !ok {
+		return nil, errors.Errorf("Could not retrieve transaction from context")
+	}
+	return transaction, nil
+}
+
+// Resolve resolves the current transaction according to the rollback flag.
+func Resolve(ctx context.Context) error {
+	transaction, err := getTransaction(ctx)
+	if err != nil {
+		global.Logger.Error(ctx, "%v", err)
+		return err
+	}
+
+	if transaction.rollback {
+		if err := rollback(ctx); err != nil {
+			global.Logger.Error(ctx, "Could not rollback transaction: %v", err)
+			return err
+		}
+		global.Logger.Info(ctx, "Rolled back transaction")
+	} else {
+		if err := commit(ctx); err != nil {
+			global.Logger.Error(ctx, "Could not commit transaction: %v", err)
+			return err
+		}
+	}
+	return nil
+}
+
+// MarkForRollback flags the transaction stored in the context for rollback.
+func MarkForRollback(ctx context.Context) {
+	transaction, err := getTransaction(ctx)
+	if err != nil {
+		global.Logger.Error(ctx, "failed to mark transaction for rollback: %v", err)
+		return
+	}
+	global.Logger.Debug(ctx, "Marked transaction for rollback.")
+	transaction.rollback = true
+}
+
+// AddPostCommitCallback adds a callback function that will be executed after the transaction is
+// successfully committed. If multiple callbacks functions are added then the order of their
+// execution isn't guaranteed, and they may run in parallel in different goroutines.
+func AddPostCommitCallback(ctx context.Context, callback func()) error {
+	transaction, err := getTransaction(ctx)
+	if err != nil {
+		return err
+	}
+	transaction.postCommitCallbacks = append(transaction.postCommitCallbacks, callback)
+	return nil
+}
+
+// AddPostRollbackCallback adds a callback function that will be executed after the transaction is
+// rolled back. If multiple callbacks functions are added then the order of their
+// execution isn't guaranteed, and they may run in parallel in different goroutines.
+func AddPostRollbackCallback(ctx context.Context, callback func()) error {
+	transaction, err := getTransaction(ctx)
+	if err != nil {
+		return err
+	}
+	transaction.postRollbackCallbacks = append(transaction.postRollbackCallbacks, callback)
+	return nil
+}
+
+// RecoverPanic recovers from a panic if one such occurred.
+func RecoverPanic(ctx context.Context, failure interface{}) {
```

Comment:
This should be a method of the `Manager` type. And it should not swallow the panic; it should instead re-panic so that other components can also process that panic should they want.

Reply:
The issue is that this should be called inside a `recover` block - that is so if were panicking _after_ a `recover` I dont see the user for that semantics? 

Heres an example inside the middleware:
```
func TransactionMiddleware(next http.Handler, manager *Manager) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Create a new Context with the transaction stored in it.
		u, err := manager.Begin()
		if err != nil {
			errors.SendInternalServerError(w, r)
		}
		ctx, err := IntoContext(u, r.Context())
		if err != nil {
			errors.SendInternalServerError(w, r)
			return
		}

		// Make a copy of the request context with the new context key,value stored in it.
		reqWithTx := r.WithContext(ctx)

		txWriter := TransactionResponseWriter{
			ResponseWriter: w,
			request:        reqWithTx,
			headerWritten:  false,
		}

		defer func() {
			if f := recover(); f != nil {
				manager.RecoverPanic(reqWithTx.Context(), f) // Here weve already panicked.
				txWriter.WriteErrIfNothingWritten(r)
			}
		}()

		// Continue handling requests.
		next.ServeHTTP(&txWriter, reqWithTx)
	})
}
```